Spectator camera system

wima12 opened the pull request:
The SpectatorCameraSystem will put the player at the new team picking screen when the game starts. After choosing team, the player will be redirected to picking class, and then to the spectator camera, showing a countdown to the next respawn wave.

The spawn system now saves which class the player picked in their spawn request. 

Clickable buttons should appear/disappear when appropriate so you can switch between screens. You can also use keyboard buttons Comma ( "," ) to swap to the class screen when you are dead, and Period ( "." ) to swap to the team picking screen. This is the defaults in DefaultConfig.

Updated the OverwatchCamera.xml so it contains the new functionality and better icons, and the NewMapWSpectatorCam.xml test map with the new overwatchCamera.

wima12 committed:
SpectatorCameraSystem handles the overwatch cameras. Set camera to a class pick screen when connecting. Cannot spawn until both team and class has been picked.

Preparations for class aware PlayerSpawnSystem.

wima12 committed:
Merge branch 'master' into TeamPickCamera

# Conflicts:
#	src/Game/Game.cpp

wima12 committed:
Moved buttons and fixed so you can't spawn with no class.

wima12 committed:
Merge branch 'Weapons' into TeamPickCamera

wima12 committed:
Unlocks mouse when changing to SpectatorCamera and locking when respawning.

wima12 committed:
Only set spectator camera if player haven't spawned already. Don't spawn player if they switch to class pick screen.

wima12 committed:
Merge branch 'master' into TeamPickCamera

# Conflicts:
#	src/Game/Systems/PlayerDeathSystem.cpp

wima12 committed:
Removed debug logging. PlayerID in InputCommands from buttons is corrected.

wima12 committed:
Replaced test textures for class picking with more proper icons.

wima12 committed:
Added clickable buttons to switch to class/team screens, as a better alternative to Comma/Period on the keyboard.

verysecrethero committed:
Merge remote-tracking branch 'origin/master' into TeamPickCamera

wima12 committed:
Don't force to  spectator camera if player presses pickTeam/Class prematurely. Players team should be remembered after picked, spawn request is not erased anymore.

wima12 committed:
Merge branch 'TeamPickCamera' of github.com:teamfisk/TacticalZ into TeamPickCamera

Tleety commented on src/Game/Systems/PlayerDeathSystem.cpp:
```diff
@@ -33,10 +35,10 @@ void PlayerDeathSystem::createDeathEffect(EntityWrapper player)
-        return;
-    }
-    if (!playerModel.HasComponent("Model") || !playerModel.HasComponent("Animation")) {
+    if (!playerModel.Valid() || !playerModel.HasComponent("Model") || !playerModel.HasComponent("Animation")) {
```
I dont think you need !playerModel.Valid() since if it has model/animation, it will be valid.

Tleety commented on src/Game/Systems/PlayerDeathSystem.cpp:
```diff
@@ -68,14 +70,35 @@ bool PlayerDeathSystem::OnEntityDeleted(Events::EntityDeleted& e)
     EntityWrapper spectatorCam = m_World->GetFirstEntityByName("SpectatorCamera");
-    if (!spectatorCam.Valid() || !spectatorCam.HasComponent("Camera") || LocalPlayer.Valid()) {
-        return false;
+    if (!spectatorCam.Valid() || !spectatorCam.HasComponent("Camera")) {
```
Same here, if it has a cameracomponent, it must be valid.

Tleety commented on src/Game/Systems/PlayerSpawnSystem.cpp:
```diff
@@ -136,27 +134,38 @@ bool PlayerSpawnSystem::OnInputCommand(Events::InputCommand& e)
-        req.Team = (ComponentInfo::EnumType)e.Value;
+        if (e.Command == "PickTeam") {
+            req.Team = (ComponentInfo::EnumType)e.Value;
+            req.Class = -1;         // TODO: -1 Signifies no class picked, enum here later?
```
Fix -1 into either a define or enum.

Tleety commented on src/Game/Systems/PlayerSpawnSystem.cpp:
```diff
@@ -136,27 +134,38 @@ bool PlayerSpawnSystem::OnInputCommand(Events::InputCommand& e)
+            // Should never get here, since you should have picked a team before you ever get a chance to pick class.
+            LOG_WARNING("Sequence error: Should not be able to pick class before team");
+            req.Team = 1;           // TODO: 1 Signifies spectator, should probably have real enum here later.
+            req.Class = (ComponentInfo::EnumType)e.Value;
```
If you somehow end up here, will you spawn as a [Class] spectator? If so this

> // A spectator can't die anyway
>       if ((ComponentInfo::EnumType)cTeam["Team"] == cTeam["Team"].Enum("Spectator")) {
>           return false;
>       }

might be a bad check?

Tleety commented on src/Game/Systems/PlayerSpawnSystem.cpp:
```diff
@@ -230,6 +241,14 @@ bool PlayerSpawnSystem::OnPlayerDeath(Events::PlayerDeath& e)
+    } else if (e.Player.HasComponent("SprintAbility")) {
+        req.Class = 3;      // Sniper
+    } else {
+        req.Class = 2;      // Defender
```
What if player has no ability? Then it becomes Defender, is this intended? If not ShieldAbility could be checked for.

Tleety commented on src/Game/Systems/SpectatorCameraSystem.cpp:
```diff
@@ -0,0 +1,90 @@
+    if (!m_CamSetToTeamPick && IsClient) {
+        // Find the class pick camera and set them to it, since they need to pick a team before they can leave the screen.
+        EntityWrapper spectatorCam = m_World->GetFirstEntityByName("PickTeamCamera");
+        if (spectatorCam.Valid() && spectatorCam.HasComponent("Camera")) {
```
Valid is unnecessary since if it has a cameraComponent it must be valid.

wima12 committed:
Falsified DepthSort for SpectatorCamera buttons and CP-HUD elements.

Tleety commented on src/Game/Systems/SpectatorCameraSystem.cpp:
```diff
@@ -0,0 +1,90 @@
+    }
+    EntityWrapper spectatorCam = m_World->GetFirstEntityByName(camName);
+    // Set the camera as active, if it exists.
+    if (spectatorCam.Valid() && spectatorCam.HasComponent("Camera")) {
```
must be valid

wima12 committed:
Class enum instead of hardcoded ints.

Removed superfluous .Valid() checks.

Check for spectator spawnrequest in case they somehow has a class, so we don't get "x players were supposed to be spawned but failed" if it happens.
